From 4dd2d39f64e754b43b068905bba29f17b3077b03 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Sat, 29 Jul 2023 16:13:35 +0800
Subject: [PATCH] fix header include non-Linux

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 cmsg.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/cmsg.cpp b/cmsg.cpp
index 95db303..aa36686 100644
--- a/cmsg.cpp
+++ b/cmsg.cpp
@@ -20,7 +20,9 @@
 #include <fcntl.h>
 #include <stdlib.h>
 #include <sys/socket.h>
+#if 0
 #include <sys/user.h>
+#endif
 
 #include <memory>
 
-- 
2.41.0

