From 48bd561bc79bae4df24297968a4e841f036e3359 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Tue, 9 Jan 2024 23:21:57 +0800
Subject: [PATCH] file: stub GetExecutablePath for non-Linux

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 file.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/file.cpp b/file.cpp
index bc831e3..a0c694d 100644
--- a/file.cpp
+++ b/file.cpp
@@ -489,7 +489,7 @@ std::string GetExecutablePath() {
   path[PATH_MAX - 1] = 0;
   return path;
 #else
-#error unknown OS
+  abort();
 #endif
 }
 
-- 
2.43.0

