From 3c9f716e99568872ea7861421c2e233d06d4a492 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Sat, 29 Jul 2023 16:16:47 +0800
Subject: [PATCH] Cygwin: find executable path like on Linux

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 file.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/file.cpp b/file.cpp
index 69ee69f..a3d2255 100644
--- a/file.cpp
+++ b/file.cpp
@@ -468,7 +468,7 @@ bool Realpath(const std::string& path, std::string* result) {
 #endif
 
 std::string GetExecutablePath() {
-#if defined(__linux__)
+#if defined(__linux__) || defined(__CYGWIN__)
   std::string path;
   android::base::Readlink("/proc/self/exe", &path);
   return path;
-- 
2.41.0

