From 2348cbcef7be34ac0603357332b3938bd1466c02 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Mon, 20 Nov 2023 22:09:23 +0800
Subject: [PATCH] mingw: use fstat stub from winsup

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 file.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/file.cpp b/file.cpp
index a3d2255..723647a 100644
--- a/file.cpp
+++ b/file.cpp
@@ -39,8 +39,10 @@
 #if defined(_WIN32)
 #include <direct.h>
 #include <windows.h>
+#include <winsup/stat_compat.h>
 #define O_NOFOLLOW 0
 #define OS_PATH_SEPARATOR '\\'
+#define fstat	_fstat_stub
 #else
 #define OS_PATH_SEPARATOR '/'
 #endif
-- 
2.42.0

