From ec3c9a9b3a977e8fa26029f2a39a79ffbc8d21e3 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Wed, 22 Mar 2023 09:24:04 +0800
Subject: [PATCH] adapt to MinGW

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 getline.c | 10 ++++++++++
 getline.h |  6 ++++++
 2 files changed, 16 insertions(+)

diff --git a/getline.c b/getline.c
index 6c3f699..5d7f4b6 100644
--- a/getline.c
+++ b/getline.c
@@ -3,9 +3,48 @@
 #include <stdlib.h>
 #include <errno.h>
 
-// MSVC specific implementation
+#ifdef _UCRT
+#ifndef WIN32_LEAN_AND_MEAN
+#define WIN32_LEAN_AND_MEAN
+#endif
+#include <Windows.h>
+
+#ifndef _IOERROR
+#define _IOERROR	(0x0010)
+#endif
+
+struct __crt_stdio_stream_data
+{
+    union
+    {
+        FILE  _public_file;
+        char* _ptr;
+    };
+
+    char*            _base;
+    int              _cnt;
+    long             _flags;
+    long             _file;
+    int              _charbuf;
+    int              _bufsiz;
+    char*            _tmpfname;
+    CRITICAL_SECTION _lock;
+};
+#endif
+
 static void fseterr(FILE *fp)
 {
+#ifdef _UCRT
+    // UCRT specific implementation
+    _lock_file(fp);
+    _InterlockedOr(&(((struct __crt_stdio_stream_data *)fp)->_flags), _IOERROR);
+    _unlock_file(fp);
+#else /* _UCRT */
+#ifdef __MINGW64__
+    // MinGW specific implementation
+    ((struct _iobuf *)fp)->_flag |= _IOERR;
+#else /* __MINGW64__ */
+    // MSVC specific implementation
     struct file { // Undocumented implementation detail
         unsigned char *_ptr;
         unsigned char *_base;
@@ -18,6 +56,8 @@ static void fseterr(FILE *fp)
     #define _IOERR 0x10
 
     ((struct file *)fp)->_flag |= _IOERR;
+#endif /* __MINGW64__ */
+#endif /* _UCRT */
 }
 
 ssize_t getdelim(char **restrict lineptr, size_t *restrict n, int delim, FILE *restrict stream)
@@ -66,7 +106,11 @@ ssize_t getdelim(char **restrict lineptr, size_t *restrict n, int delim, FILE *r
     return nread;
 }
 
+#ifdef __MINGW64__
+ssize_t msvc_getline(char **restrict lineptr, size_t *restrict n, FILE *restrict stream)
+#else
 ssize_t getline(char **restrict lineptr, size_t *restrict n, FILE *restrict stream)
+#endif
 {
     return getdelim(lineptr, n, '\n', stream);
 }
diff --git a/getline.h b/getline.h
index ffbb699..b7bab62 100644
--- a/getline.h
+++ b/getline.h
@@ -4,9 +4,15 @@
 #include <stdio.h>
 
 #define restrict __restrict
+#ifndef __MINGW64__
 typedef long long ssize_t;
+#endif
 
+#ifdef __MINGW64__
+ssize_t msvc_getline(char **restrict lineptr, size_t *restrict n, FILE *restrict stream);
+#else
 ssize_t getline(char **restrict lineptr, size_t *restrict n, FILE *restrict stream);
+#endif
 ssize_t getdelim(char **restrict lineptr, size_t *restrict n, int delim, FILE *restrict stream);
 
 /*
-- 
2.40.0.windows.1

