From 804f8c7b9e13b20ea8e7ac9002f7a6a9f43a5ca0 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Thu, 9 Mar 2023 20:34:50 +0800
Subject: [PATCH] native: base: rust: add errno_location other platforms

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/base/misc.rs | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/native/src/base/misc.rs b/native/src/base/misc.rs
index 521acb6..569cf88 100644
--- a/native/src/base/misc.rs
+++ b/native/src/base/misc.rs
@@ -115,8 +117,25 @@ pub fn ptr_to_str<'a, T>(ptr: *const T) -> &'a str {
     }
 }
 
+#[cfg(all(target_os = "windows", target_env = "gnu"))]
+extern "cdecl" {
+    pub fn _errno() -> *mut libc::c_int;
+}
+
 pub fn errno() -> &'static mut i32 {
-    unsafe { &mut *libc::__errno() }
+    cfg_if::cfg_if! {
+        if #[cfg(any(target_os = "android", target_os = "cygwin"))] {
+            unsafe { &mut *libc::__errno() }
+        } else if #[cfg(target_os = "linux")] {
+            unsafe { &mut *libc::__errno_location() }
+        } else if #[cfg(target_vendor = "apple")] {
+            unsafe { &mut *libc::__error() }
+        } else if #[cfg(all(target_os = "windows", target_env = "gnu"))] {
+            unsafe { &mut *_errno() }
+        } else {
+            compile_error!("unknown errno location");
+        }
+    }
 }
 
 pub fn error_str() -> &'static str {
-- 
2.40.0

