From e9b79ff55ce85478c37f5d71361f058cf384df6c Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Thu, 9 Mar 2023 23:26:52 +0800
Subject: [PATCH 2/5] native: base: disable missing.hpp

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/base/include/base.hpp  | 2 ++
 1 files changed, 2 insertions(+)

diff --git a/native/src/base/include/base.hpp b/native/src/base/include/base.hpp
index a8481f3..e1b588e 100644
--- a/native/src/base/include/base.hpp
+++ b/native/src/base/include/base.hpp
@@ -1,6 +1,8 @@
 #pragma once
 
+#if 0
 #include "../missing.hpp"
+#endif
 #include "../xwrap.hpp"
 #include "../files.hpp"
 #include "../misc.hpp"
-- 
2.39.2

