From 44746641c2db4789670b461c3ebe43a7ddb00431 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Thu, 9 Mar 2023 22:21:43 +0800
Subject: [PATCH 4/5] native: base: stream: replace fpos_t with off_t on Linux

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/base/stream.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/native/src/base/stream.cpp b/native/src/base/stream.cpp
index 1bc7837..7bd7544 100644
--- a/native/src/base/stream.cpp
+++ b/native/src/base/stream.cpp
@@ -18,7 +18,11 @@ static int strm_write(void *v, const char *buf, int len) {
     return len;
 }
 
+#ifdef __ANDROID__
 static fpos_t strm_seek(void *v, fpos_t off, int whence) {
+#else
+static off_t strm_seek(void *v, off_t off, int whence) {
+#endif
     auto strm = static_cast<channel *>(v);
     return strm->seek(off, whence);
 }
-- 
2.39.2

