From 16adfae16a805b1fd0c7a59fe68fa6044ecfc027 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Mon, 20 Mar 2023 01:43:44 +0800
Subject: [PATCH 3/6] base: misc: add partial pdeathsig impl for non-Linux
 platforms

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/base/misc.cpp | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/native/src/base/misc.cpp b/native/src/base/misc.cpp
index 9a5b6d7..cf16b3b 100644
--- a/native/src/base/misc.cpp
+++ b/native/src/base/misc.cpp
@@ -39,11 +39,39 @@ int fork_dont_care() {
     return 0;
 }
 
+#ifndef __linux__
+static int n_childern = 0;
+static int *child_pids = NULL;
+static int pdeathsig_regged = 0;
+
+void kill_childern() {
+    int i;
+
+    for (i = 0; i < n_childern; i++) {
+    }
+
+    if (child_pids)
+        free(child_pids);
+}
+#endif
+
 int fork_no_orphan() {
     int pid = xfork();
+#ifdef __linux__
     if (pid)
         return pid;
     prctl(PR_SET_PDEATHSIG, SIGKILL);
+#else
+    if (!pdeathsig_regged)
+        std::atexit(kill_childern);
+    if (pid) {
+        int child_pid_idx = n_childern;
+        child_pids = (int *) realloc(child_pids, ++n_childern * sizeof(int));
+        child_pids[child_pid_idx] = pid;
+        return pid;
+    } else if (n_childern)
+        n_childern = 0;
+#endif
     if (getppid() == 1)
         exit(1);
     return 0;
-- 
2.40.0

