From 14e17ecef680586e216e0a61ae567ebdbd973a38 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Mon, 20 Mar 2023 01:47:37 +0800
Subject: [PATCH 5/6] base: selinux: stub setcon on non-Linux platforms

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/base/selinux.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/native/src/base/selinux.cpp b/native/src/base/selinux.cpp
index 269d35f..081219b 100644
--- a/native/src/base/selinux.cpp
+++ b/native/src/base/selinux.cpp
@@ -39,6 +39,7 @@ static void __freecon(char *s) {
 }
 
 static int __setcon(const char *ctx) {
+#ifdef __linux__
     int fd = open("/proc/self/attr/current", O_WRONLY | O_CLOEXEC);
     if (fd < 0)
         return fd;
@@ -46,6 +47,9 @@ static int __setcon(const char *ctx) {
     int rc = write(fd, ctx, len);
     close(fd);
     return rc != len;
+#else
+    return 0;
+#endif
 }
 
 static int __getfilecon(const char *path, char **ctx) {
-- 
2.40.0

