From 3e45865569fa86ec5f3c0c9d9590debb83cedca4 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Mon, 20 Mar 2023 03:51:40 +0800
Subject: [PATCH 2/2] base: rust: find proc fds from /dev/fd on macOS

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/base/files.rs | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/native/src/base/files.rs b/native/src/base/files.rs
index 9752f1e..1ceeebd 100644
--- a/native/src/base/files.rs
+++ b/native/src/base/files.rs
@@ -80,12 +80,20 @@ pub fn readlink(path: &CStr, data: &mut [u8]) -> isize {
     unsafe { unsafe_impl::readlink(path.as_ptr(), data.as_mut_ptr(), data.len()) }
 }
 
+#[cfg(target_os = "linux")]
 pub fn fd_path(fd: RawFd, buf: &mut [u8]) -> isize {
     let mut fd_buf: [u8; 40] = [0; 40];
     let fd_path = bfmt_cstr!(&mut fd_buf, "/proc/self/fd/{}", fd);
     readlink(fd_path, buf)
 }
 
+#[cfg(target_vendor = "apple")]
+pub fn fd_path(fd: RawFd, buf: &mut [u8]) -> isize {
+    let mut fd_buf: [u8; 40] = [0; 40];
+    let fd_path = bfmt_cstr!(&mut fd_buf, "/dev/fd/{}", fd);
+    readlink(fd_path, buf)
+}
+
 // Inspired by https://android.googlesource.com/platform/bionic/+/master/libc/bionic/realpath.cpp
 pub fn realpath(path: &CStr, buf: &mut [u8]) -> isize {
     if let Some(fd) = open_fd!(path, O_PATH | O_CLOEXEC) {
-- 
2.40.0

