From e266ed361a2d5baa8215ccdac7c6be744503b96f Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Mon, 20 Mar 2023 04:30:03 +0800
Subject: [PATCH] base: xwrap: disable dup3 on non-Linux

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/base/xwrap.rs  | 6 +++++-
 1 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/native/src/base/xwrap.rs b/native/src/base/xwrap.rs
index 08a81c6..a0f560f 100644
--- a/native/src/base/xwrap.rs
+++ b/native/src/base/xwrap.rs
@@ -461,6 +464,7 @@ extern "C" fn xdup2(oldfd: RawFd, newfd: RawFd) -> RawFd {
 }
 
 #[no_mangle]
+#[cfg(target_os = "linux")]
 extern "C" fn xdup3(oldfd: RawFd, newfd: RawFd, flags: i32) -> RawFd {
     unsafe {
         let fd = libc::syscall(SYS_dup3, oldfd, newfd, flags) as RawFd;
-- 
2.40.0

