From 20229cb0f5eb8a0b07538fd89b1ae7435e9de8ce Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Mon, 20 Mar 2023 04:39:06 +0800
Subject: [PATCH] base: xwrap: rust: alias lseek to lseek64 on non-Linux system

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/base/xwrap.rs | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/native/src/base/xwrap.rs b/native/src/base/xwrap.rs
index a0f560f..83129a5 100644
--- a/native/src/base/xwrap.rs
+++ b/native/src/base/xwrap.rs
@@ -10,6 +10,8 @@ use libc::{
 #[cfg(target_os = "linux")]
 use libc::SYS_dup3;
 
+use cfg_if::cfg_if;
+
 use crate::{cstr, errno, error, mkdirs, perror, ptr_to_str, realpath};
 
 mod unsafe_impl {
@@ -218,7 +220,13 @@ pub fn xxread(fd: RawFd, data: &mut [u8]) -> isize {
 #[no_mangle]
 pub extern "C" fn xlseek64(fd: RawFd, offset: i64, whence: i32) -> i64 {
     unsafe {
-        let r = libc::lseek64(fd, offset, whence);
+        cfg_if! {
+            if #[cfg(target_os = "linux")] {
+                let r = libc::lseek64(fd, offset, whence);
+            } else {
+                let r = libc::lseek(fd, offset, whence);
+            }
+        }
         if r < 0 {
             perror!("lseek64");
         }
-- 
2.40.0

