From 68274f3c90a9a3f30b653f728ab8b77095641010 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Mon, 20 Mar 2023 05:38:36 +0800
Subject: [PATCH] base: rust: xwrap: disable more function on non-Linux

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/base/xwrap.rs | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/native/src/base/xwrap.rs b/native/src/base/xwrap.rs
index 583474d..9bf814b 100644
--- a/native/src/base/xwrap.rs
+++ b/native/src/base/xwrap.rs
@@ -91,6 +91,7 @@ mod unsafe_impl {
     }
 
     #[no_mangle]
+    #[cfg(any(target_os = "linux", target_os = "android"))]
     pub unsafe extern "C" fn xsendfile(
         out_fd: RawFd,
         in_fd: RawFd,
@@ -246,6 +247,7 @@ pub fn xpipe2(fds: &mut [i32; 2], flags: i32) -> i32 {
 }
 
 #[no_mangle]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 pub extern "C" fn xsetns(fd: RawFd, nstype: i32) -> i32 {
     unsafe {
         let r = libc::setns(fd, nstype);
@@ -257,6 +259,7 @@ pub extern "C" fn xsetns(fd: RawFd, nstype: i32) -> i32 {
 }
 
 #[no_mangle]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 pub extern "C" fn xunshare(flags: i32) -> i32 {
     unsafe {
         let r = libc::unshare(flags);
@@ -357,6 +360,7 @@ pub extern "C" fn xlisten(socket: i32, backlog: i32) -> i32 {
 }
 
 #[no_mangle]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 pub extern "C" fn xaccept4(
     sockfd: RawFd,
     addr: *mut sockaddr,
@@ -558,6 +562,7 @@ pub extern "C" fn xlinkat(
 }
 
 #[no_mangle]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 pub extern "C" fn xmount(
     src: *const c_char,
     target: *const c_char,
@@ -575,6 +580,7 @@ pub extern "C" fn xmount(
 }
 
 #[no_mangle]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 pub extern "C" fn xumount(target: *const c_char) -> i32 {
     unsafe {
         let r = libc::umount(target);
@@ -586,6 +592,7 @@ pub extern "C" fn xumount(target: *const c_char) -> i32 {
 }
 
 #[no_mangle]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 pub extern "C" fn xumount2(target: *const c_char, flags: i32) -> i32 {
     unsafe {
         let r = libc::umount2(target, flags);
@@ -639,6 +646,7 @@ pub extern "C" fn xmkdirat(dirfd: RawFd, path: *const c_char, mode: mode_t) -> i
 }
 
 #[inline]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 pub fn xsendfile(out_fd: RawFd, in_fd: RawFd, offset: Option<&mut off_t>, count: usize) -> isize {
     unsafe {
         let p = offset.map_or(ptr::null_mut(), |it| it);
-- 
2.40.0

