From bc292eb6b6cda6ac88c4f756b0feaa470c50494b Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Fri, 28 Jul 2023 09:16:47 +0800
Subject: [PATCH] fix MetadataExt import on other OS

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/base/files.rs | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/native/src/base/files.rs b/native/src/base/files.rs
index bcaea46..196c000 100644
--- a/native/src/base/files.rs
+++ b/native/src/base/files.rs
@@ -4,7 +4,22 @@ use std::ffi::CStr;
 use std::fs::File;
 use std::io::{BufRead, Read, Seek, SeekFrom, Write};
 use std::ops::Deref;
+
+#[cfg(target_os = "android")]
 use std::os::android::fs::MetadataExt;
+
+#[cfg(target_os = "linux")]
+use std::os::linux::fs::MetadataExt;
+
+#[cfg(target_vendor = "apple")]
+use std::os::macos::fs::MetadataExt;
+
+#[cfg(target_os = "windows")]
+use std::os::windows::fs::MetadataExt;
+
+#[cfg(target_os = "cygwin")]
+use std::os::cygwin::fs::MetadataExt;
+
 use std::os::fd::{AsFd, BorrowedFd, IntoRawFd};
 use std::os::unix::fs::FileTypeExt;
 use std::os::unix::io::{AsRawFd, FromRawFd, OwnedFd, RawFd};
-- 
2.41.0

