From 8641e93ec645f2db20e74aafb7f3fddc4b7a2537 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Fri, 28 Jul 2023 12:33:12 +0800
Subject: [PATCH] macOS: makedev is unsafe func

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/boot/cpio.rs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/native/src/boot/cpio.rs b/native/src/boot/cpio.rs
index d19ce29..0e36b4d 100644
--- a/native/src/boot/cpio.rs
+++ b/native/src/boot/cpio.rs
@@ -353,8 +353,8 @@ impl Cpio {
                 symlink(Path::new(&std::str::from_utf8(entry.data.as_slice())?), out)?;
             }
             S_IFBLK | S_IFCHR => {
-                let dev = makedev(entry.rdevmajor.try_into()?, entry.rdevminor.try_into()?);
                 unsafe {
+                    let dev = makedev(entry.rdevmajor.try_into()?, entry.rdevminor.try_into()?);
                     mknod(
                         out.to_str().unwrap().as_ptr() as *const c_char,
                         entry.mode,
-- 
2.41.0

