From 90fd60db826362081ccd0802775faa0cd031e587 Mon Sep 17 00:00:00 2001
From: Ookiineko <chiisaineko@protonmail.com>
Date: Sat, 29 Jul 2023 16:30:25 +0800
Subject: [PATCH] disable more functions on non-linux (2/2)

Signed-off-by: Ookiineko <chiisaineko@protonmail.com>
---
 native/src/base/misc.cpp  | 2 ++
 1 files changed, 2 insertions(+)

diff --git a/native/src/base/misc.cpp b/native/src/base/misc.cpp
index fb7d7ca..903da92 100644
--- a/native/src/base/misc.cpp
+++ b/native/src/base/misc.cpp
@@ -270,6 +270,7 @@ uint32_t binary_gcd(uint32_t u, uint32_t v) {
     return u << shift;
 }
 
+#ifdef __linux__
 int switch_mnt_ns(int pid) {
     char mnt[32];
     ssprintf(mnt, sizeof(mnt), "/proc/%d/ns/mnt", pid);
@@ -283,6 +284,7 @@ int switch_mnt_ns(int pid) {
     close(fd);
     return ret;
 }
+#endif
 
 string &replace_all(string &str, string_view from, string_view to) {
     size_t pos = 0;
-- 
2.41.0

